gemstone/src/swapper/mayan/forwarder.rs

```diff
@@ -0,0 +1,126 @@
+        token_in: &str,
+        amount_in: U256,
+        permit: Option<MayanSwiftPermit>,
+        mayan_protocol: &str,
+        protocol_data: Vec<u8>,
+    ) -> Result<Vec<u8>, MayanForwarderError> {
+        let call_data = IMayanForwarder::forwardERC20Call {
+            tokenIn: Address::from_str(token_in).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid token address: {}", e),
+            })?,
+            amountIn: amount_in,
+            permitParams: permit.map_or(MayanSwiftPermit::zero().to_contract_params(), |p| p.to_contract_params()),
+            mayanProtocol: Address::from_str(mayan_protocol).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid protocol address: {}", e),
+            })?,
+            protocolData: protocol_data.into(),
+        }
+        .abi_encode();
+
+        Ok(call_data)
+    }
+
+    pub async fn encode_swap_and_forward_eth_call(
+        &self,
+        amount_in: U256,
+        swap_protocol: &str,
+        swap_data: Vec<u8>,
+        middle_token: &str,
+        min_middle_amount: U256,
+        mayan_protocol: &str,
+        mayan_data: Vec<u8>,
+    ) -> Result<Vec<u8>, MayanForwarderError> {
+        let call_data = IMayanForwarder::swapAndForwardEthCall {
+            amountIn: amount_in,
+            swapProtocol: Address::from_str(swap_protocol).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid swap protocol address: {}", e),
+            })?,
+            swapData: swap_data.into(),
+            middleToken: Address::from_str(middle_token).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid middle token address: {}", e),
+            })?,
+            minMiddleAmount: min_middle_amount,
+            mayanProtocol: Address::from_str(mayan_protocol).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid mayan protocol address: {}", e),
+            })?,
+            mayanData: mayan_data.into(),
+        }
+        .abi_encode();
+
+        Ok(call_data)
+    }
+
+    pub fn encode_swap_and_forward_erc20_call(
+        &self,
+        token_in: &str,
+        amount_in: U256,
+        permit: Option<MayanSwiftPermit>,
+        swap_protocol: &str,
+        swap_data: Vec<u8>,
+        middle_token: &str,
+        min_middle_amount: U256,
+        mayan_protocol: &str,
+        mayan_data: Vec<u8>,
+    ) -> Result<Vec<u8>, MayanForwarderError> {
+        let call_data = IMayanForwarder::swapAndForwardERC20Call {
+            tokenIn: Address::from_str(token_in).map_err(|e| MayanForwarderError::ABIError {
+                msg: format!("Invalid token address: {}", e),
+            })?,
+            amountIn: amount_in,
+            permitParams: permit.map_or(MayanSwiftPermit::zero().to_contract_params(), |p| p.to_contract_params()),
+            swapProtocol: Address::from_str(swap_protocol).map_err(|e| MayanForwarderError::ABIError {
```

Comment:
if you add a `From<AddressError> for MayanForwarderError`, it would be easier to return an address error just passing `.map_err(MayanForwarderError::from`